### PR TITLE
Introduce the /priceSubscriptions endpoint & support for Subscription creation & edit pricing

### PR DESCRIPTION
**Pull request by ckittrell**
https://github.com/BroadleafCommerce/MicroPM/issues/5135

### CODE REVIEW

File: services/src/main/java/com/broadleafcommerce/subscriptionoperation/web/endpoint/SubscriptionPricingEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+@RequiredArgsConstructor
+@FrameworkRestController
+public class SubscriptionPricingEndpoint {
+
+    @Getter(AccessLevel.PROTECTED)
+    protected final SubscriptionPricingService subscriptionPricingService;
+
+    @FrameworkPostMapping(value = "/price-subscriptions")
+    // TODO
```
Comment (ckittrell) [inline services/src/main/java/com/broadleafcommerce/subscriptionoperation/web/endpoint/SubscriptionPricingEndpoint.java:44]: TODO!

File: services/src/main/java/com/broadleafcommerce/subscriptionoperation/domain/enums/DefaultTermDurationType.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+import com.broadleafcommerce.order.common.domain.RecurringPriceDetail;
+
+/**
+ * Defines the default term duration types possible for {@link RecurringPriceDetail recurring
+ * prices}.
+ *
+ * @author Nathan Moore (nathandmoore)
+ * @since Pricing Client 2.1.0
+ *
```
Comment (nathandmoore) [inline services/src/main/java/com/broadleafcommerce/subscriptionoperation/domain/enums/DefaultTermDurationType.java:27]:
```suggestion
```

File: services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java [hunk 3]
```diff
@@ -0,0 +1,875 @@
...
+
+        if (DefaultTermDurationType.isDays(termDurationType)) {
+            ldt.plusDays(termDurationLength);
+        } else if (DefaultTermDurationType.isWeeks(termDurationType)) {
+            ldt.plusWeeks(termDurationLength);
+        } else if (DefaultTermDurationType.isMonths(termDurationType)) {
+            ldt.plusMonths(termDurationLength);
+        } else if (DefaultTermDurationType.isYears(termDurationType)) {
+            ldt.plusYears(termDurationLength);
+        }
```
Comment (nathandmoore) [inline services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java:257]: Javadocs say
> Returns a copy of this {@\code LocalDateTime} with the specified number of days added.
> ...
> This instance is immutable and unaffected by this method call.

So pretty sure you need to reassign `ltd` to the result

```suggestion
        if (DefaultTermDurationType.isDays(termDurationType)) {
             ltd = ldt.plusDays(termDurationLength);
        } else if (DefaultTermDurationType.isWeeks(termDurationType)) {
             ltd = ldt.plusWeeks(termDurationLength);
        } else if (DefaultTermDurationType.isMonths(termDurationType)) {
            ldt.plusMonths(termDurationLength);
        } else if (DefaultTermDurationType.isYears(termDurationType)) {
            ltd = ldt.plusYears(termDurationLength);
        }
```

File: services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java [hunk 4]
```diff
@@ -0,0 +1,875 @@
...
+
+        return subscriptionProvider.readSubscriptionById(existingSubscriptionId, contextInfo);
+    }
+
+    protected String getExistingSubscriptionItemId(@lombok.NonNull CartItem cartItem,
+            @Nullable ContextInfo contextInfo) {
+        return MapUtils.getString(cartItem.getInternalAttributes(), EXISTING_SUBSCRIPTION_ITEM_ID);
+    }
+
+    protected void populateFromExistingSubscription(
```
Comment (nathandmoore) [inline services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java:186]:
```suggestion
    protected void populatePricingContextFromExistingSubscription(
```

File: services/src/test/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingServiceCreateFlowTest.java [hunk 5]
```diff
@@ -0,0 +1,1009 @@
...
+
+        assertThat(subscriptionPricingContext.getFlow())
+                .isEqualTo(DefaultSubscriptionActionFlow.CREATE.name());
+        assertThat(subscriptionPricingContext.getExistingSubscriptionId()).isBlank();
+        assertThat(subscriptionPricingContext.getPaymentStrategy())
+                .isEqualTo(DefaultSubscriptionPaymentStrategy.POSTPAID.name());
+        assertThat(subscriptionPricingContext.getPeriodType())
+                .isEqualTo(DefaultSubscriptionPeriodType.MONTHLY.name());
+        assertThat(subscriptionPricingContext.getPeriodFrequency()).isEqualTo(1);
+        assertThat(subscriptionPricingContext.getEndOfTermsDate()).isNotNull();
```
Comment (nathandmoore) [inline services/src/test/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingServiceCreateFlowTest.java:134]: Should verify a specific value. Something like now + 30 days, likely have to ignore time part of the instant.

File: services/src/test/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingServiceEditFlowTest.java [hunk 6]
```diff
@@ -0,0 +1,982 @@
...
+        } else if (isQuarterly(pricingContext.getPeriodType())) {
+            return 90L * pricingContext.getPeriodFrequency();
+        } else if (isAnnually(pricingContext.getPeriodType())) {
+            return 365L * pricingContext.getPeriodFrequency();
+        } else {
+            throw new IllegalArgumentException(String.format(
+                    "Unable to determine number of days in the period. Period type: %s.",
+                    pricingContext.getPeriodType()));
+        }
+    }
```
Comment (nathandmoore) [inline services/src/test/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingServiceEditFlowTest.java:936]: Seems like you are just recreating the business logic to test the business logic. If there's a mistake in the business logic, might also just copy that mistake to the test of it.

File: services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java [hunk 7]
```diff
@@ -0,0 +1,875 @@
...
+            periodDefinition.setPeriodEndDate(periodEndDate);
+
+            estimatedFuturePaymentPeriods.put(period, periodDefinition);
+            previousBillDate = periodDefinition.getBillDate();
+        }
+
+        return estimatedFuturePaymentPeriods;
+    }
+
+    protected PeriodDefinition buildAtypicalFirstPeriodDefinition(
```
Comment (nathandmoore) [inline services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java:321]: Seems more polite to say custom than atypical.
```suggestion
    protected PeriodDefinition buildCustomFirstPeriodDefinition(
```

File: services/src/test/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingServiceEditFlowTest.java (hunk 6 above)
Comment (ckittrell) [reply services/src/test/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingServiceEditFlowTest.java]: I see your point! I'll plan to refine the test suites as part of the upgrade/downgrade pricing ticket in Sprint 4.

File: services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java (hunk 7 above)
Comment (ckittrell) [reply services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java:322]: Planning to leave this one as-is

**Review by nathandmoore — APPROVED** [review]

**Pull request merged**